### PR TITLE
feat: Replay history from files

### PR DESCRIPTION
**Pull request by Sushisource**
## What was changed
Added ability to replay history from history files

## Why?
Nice to be able to verify changes to workflow code is compatible with existing histories

## Checklist
Draft until some todos are resolved, and depends on https://github.com/temporalio/sdk-core/pull/261 (CI will fail until that is merged and I update)

1. Closes https://github.com/temporalio/sdk-typescript/issues/255

2. How was this tested:
Added tests

3. Any docs updates needed?
Yes

### CODE REVIEW

File: packages/test/src/test-replay.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+});
+
+test('cancel-fake-progress-replay', async (t) => {
+  // TODO: Auto-setup of worker for these kinds of tests
+
+  const histBin = await fs.promises.readFile(
+    path.resolve(__dirname, '../src/history_files/cancel_fake_progress_history.bin')
+  );
+  const hist = History.decode(histBin);
+  await Worker.createReplay(
```
[the remark this answers is not in the document]
Comment (bergundy) [reply packages/test/src/test-replay.ts:30]: I'm trying to figure out if there's a good reason to expose `Worker.createReplay` directly.
Maybe all we want is a function to create and run which throws when there's non-determinism errors.

File: packages/worker/src/core.ts [hunk 2]
```diff
@@ -81,20 +93,59 @@ export class Core {
...
   }
 
   /**
-   * Default options get overridden when Core is installed and are remembered in case Core is
-   * re-instantiated after being shut down
+   * Get or instantiate the singleton Core object
+   *
+   * If Core has not been instantiated with {@link install} or this method,
+   * a new Core instance will be installed and configured to connect to
+   * the local docker compose setup.
```
Comment (bergundy) [inline packages/worker/src/core.ts:128]: I'd use the chance to rephrase this to:
```suggestion
   * a local server.
```

File: packages/worker/src/core.ts [hunk 3]
```diff
@@ -253,3 +260,65 @@ export class Core {
...
+  protected static instantiator?: 'install' | 'instance';
+  /**
+   * Default options get overridden when Core is installed and are remembered in case Core is
+   * re-instantiated after being shut down
+   */
+  protected static defaultOptions: CoreOptions = {};
+
+  // TODO: How do we de-dupe all this static goop? Maybe have some non-static
+  //   core singleton generic type which these hold a static instance of
+  public static async install(options: CoreOptions): Promise<ReplayCore> {
```
Comment (bergundy) [inline packages/worker/src/core.ts:275]: You can reuse the original install signature if you use this as a template param:
https://github.com/temporalio/sdk-typescript/blob/afc0d964ecf21bbe5e79b28300e46bd2273dd38e/packages/worker/src/workflow/vm.ts#L77

File: packages/worker/src/core.ts [hunk 4]
```diff
@@ -253,3 +260,65 @@ export class Core {
...
+   * Create a replay core instance, for replaying static histories
+   */
+  public static async create(options: CoreOptions): Promise<ReplayCore> {
+    const compiledOptions = this.compileOptions(options);
+    const native = await promisify(newReplayCore)(compiledOptions.telemetryOptions);
+
+    return new this(native, compiledOptions);
+  }
+
+  // TODO: accept either history or binary
```
Comment (bergundy) [inline packages/worker/src/core.ts:314]: or json, but we can leave that for another pr

File: packages/worker/src/index.ts [hunk 5]
```diff
@@ -11,7 +11,7 @@
 export { State, Worker, DataConverter, defaultDataConverter, errors } from './worker';
 export { WorkerOptions, CompiledWorkerOptions } from './worker-options';
 export { ServerOptions, TLSConfig } from './server-options';
-export { Core, CoreOptions } from './core';
+export { Core, CoreOptions, ReplayCore } from './core';
```
Comment (bergundy) [inline packages/worker/src/index.ts:14]: I'm not sure we want to export it here. Users typically shouldn't use this API

File: packages/worker/src/worker.ts [hunk 6]
```diff
@@ -186,6 +198,40 @@ export class Worker {
...
+      taskQueue: options.testName + '-' + this.replayWorkerCount,
+      workflowThreadPoolSize: 1,
+      ...options,
+    };
+    const compiledOptions = compileWorkerOptions(addDefaultWorkerOptions(fixedUpOptions));
+    const replayWorker = await nativeWorkerCtor.createReplay(compiledOptions, history);
+    this.replayWorkerCount += 1;
+    const constructedWorker = await this.bundleWorker(compiledOptions, replayWorker);
+
+    // TODO: Make running lazy? Return type which users can then explicitly start running?
```
Comment (bergundy) [inline packages/worker/src/worker.ts:219]: Don't see why this is needed

File: packages/worker/src/worker.ts [hunk 7]
```diff
@@ -186,6 +198,40 @@ export class Worker {
...
+  }
+
+  /**
+   * Create a replay Worker.
+   */
+  public static async createReplay(options: ReplayWorkerOptions, history: History): Promise<void> {
+    const nativeWorkerCtor: WorkerConstructor = this.nativeWorkerCtor;
+    const fixedUpOptions = {
+      taskQueue: options.testName + '-' + this.replayWorkerCount,
+      workflowThreadPoolSize: 1,
```
Comment (bergundy) [inline packages/worker/src/worker.ts:211]: We should instead start the worker in debug mode which doesn't create a thread pool at all.
`debugMode` should be omitted from ReplayWorkerOptions.

File: packages/worker/src/worker.ts [hunk 8]
```diff
@@ -764,10 +811,18 @@ export class Worker {
           return { activation, parentSpan };
         });
       } catch (err) {
-        parentSpan.setStatus({ code: otel.SpanStatusCode.ERROR }).end();
+        if (!(err instanceof errors.ShutdownError)) {
```
Comment (bergundy) [inline packages/worker/src/worker.ts:814]: Looks like [instrument](https://github.com/temporalio/sdk-typescript/blob/ce23c2c96b3e9bbaa46378d5658f6698d3f6bd5f/packages/worker/src/tracing.ts#L28) already marks the span with code: ERROR

File: packages/worker/src/worker.ts [hunk 9]
```diff
@@ -764,10 +811,18 @@ export class Worker {
...
+          parentSpan.setStatus({ code: otel.SpanStatusCode.ERROR }).end();
+        }
         throw err;
       }
-    });
+    }).pipe(
+      tap({
+        complete: () => {
+          this.workflowPollerStateSubject.next('SHUTDOWN');
```
Comment (bergundy) [inline packages/worker/src/worker.ts:822]: We should change the state to `FAILED` on error here and perhaps even flush the cache when that state is reached.

File: packages/worker/src/worker.ts [hunk 10]
```diff
@@ -908,7 +970,7 @@ export class Worker {
    */
   async run(): Promise<void> {
     if (this.state !== 'INITIALIZED') {
-      throw new IllegalStateError('Poller was aleady started');
+      throw new IllegalStateError('Poller was already started');
```
Comment (bergundy) [inline packages/worker/src/worker.ts:973]: There are already enough "r"s in this sentence.

File: packages/core-bridge/Cargo.toml [hunk 11]
```diff
@@ -19,7 +19,11 @@ prost = "0.9"
...
-temporal-sdk-core-api = { version = "0.1.0", path = "./sdk-core/core-api" }
-temporal-sdk-core-protos = { version = "0.1.0", path = "./sdk-core/sdk-core-protos" }
+#temporal-sdk-core = { version = "*", path = "./sdk-core/core" }
+#temporal-sdk-core-api = { version = "*", path = "./sdk-core/core-api" }
+#temporal-sdk-core-protos = { version = "*", path = "./sdk-core/sdk-core-protos" }
+
+temporal-sdk-core = { version = "*", path = "../../../sdk-core/core" }
+temporal-sdk-core-api = { version = "*", path = "../../../sdk-core/core-api" }
+temporal-sdk-core-protos = { version = "*", path = "../../../sdk-core/sdk-core-protos" }
+temporal-sdk-core-test-utils = { version = "*", path = "../../../sdk-core/test-utils" }
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply packages/core-bridge/Cargo.toml:29]: Fixed

File: packages/core-bridge/Cargo.toml [hunk 12]
```diff
@@ -19,7 +19,11 @@ prost = "0.9"
...
-temporal-sdk-core = { version = "0.1.0-alpha.1", path = "./sdk-core/core" }
-temporal-sdk-core-api = { version = "0.1.0", path = "./sdk-core/core-api" }
-temporal-sdk-core-protos = { version = "0.1.0", path = "./sdk-core/sdk-core-protos" }
+#temporal-sdk-core = { version = "*", path = "./sdk-core/core" }
+#temporal-sdk-core-api = { version = "*", path = "./sdk-core/core-api" }
+#temporal-sdk-core-protos = { version = "*", path = "./sdk-core/sdk-core-protos" }
+
+temporal-sdk-core = { version = "*", path = "../../../sdk-core/core" }
+temporal-sdk-core-api = { version = "*", path = "../../../sdk-core/core-api" }
+temporal-sdk-core-protos = { version = "*", path = "../../../sdk-core/sdk-core-protos" }
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply packages/core-bridge/Cargo.toml:28]: Makes sense, done.

File: packages/test/src/mock-native-worker.ts [hunk 13]
```diff
@@ -52,6 +52,10 @@ export class MockNativeWorker implements NativeWorkerLike {
     return new this();
   }
 
+  public static async createReplay(): Promise<NativeWorkerLike> {
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply packages/test/src/mock-native-worker.ts:55]: Because otherwise it doesn't adhere to the `WorkerConstructor` interface

File: packages/test/src/test-replay.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sushisource) [reply packages/test/src/test-replay.ts:30]: I think that's typically what people want. We could just lazily init a replay core that is always used for all replay workers.

File: packages/worker/src/index.ts (hunk 5 above)
Comment (Sushisource) [reply packages/worker/src/index.ts:14]: It's needed if you want to, for some reason, control your own replay core instance and register workers against it directly. I think for now we can do the lazy-initted core and just use that unless we later discover a reason not to

Comment (Sushisource) [reply packages/worker/src/index.ts:14]: Oh, no there is a good reason, so you can control logging options when desired.

File: packages/worker/src/worker.ts (hunk 8 above)
Comment (Sushisource) [reply packages/worker/src/worker.ts:814]: So are you suggesting we invert this to un-set it?

File: packages/worker/src/worker.ts (hunk 9 above)
Comment (Sushisource) [reply packages/worker/src/worker.ts:822]: Seems like just falling through to https://github.com/temporalio/sdk-typescript/blob/ce23c2c96b3e9bbaa46378d5658f6698d3f6bd5f/packages/worker/src/worker.ts#L996 is the right place to handle that? I think cache flushing subject to that probably goes in another PR.

File: packages/worker/src/core.ts (hunk 3 above)
Comment (Sushisource) [reply packages/worker/src/core.ts:275]: I don't think that works - not without making the entire class generic at least, because there is still only one static `_instance` field and it starts being nonassignable when you do this.

Comment (Sushisource) [reply packages/worker/src/core.ts:275]: Actually I was able to overcome that but it still doesn't work. The static `install` on the base class will use the base class' `_instance` field, not the subclass, so we end up erroneously constructing a real core.

File: packages/worker/src/rxutils.ts [hunk 14]
```diff
@@ -1,5 +1,5 @@
 import { GroupedObservable, OperatorFunction, ObservableInput, pipe, Subject } from 'rxjs';
-import { groupBy, map, scan, mergeScan } from 'rxjs/operators';
```
Comment (bergundy) [inline packages/worker/src/rxutils.ts:2]: scan hater

File: packages/worker/src/worker.ts [hunk 15]
```diff
@@ -186,6 +198,43 @@ export class Worker {
...
+      ...options,
+    };
+    this.replayWorkerCount += 1;
+    const compiledOptions = compileWorkerOptions(addDefaultWorkerOptions(fixedUpOptions));
+    const replayWorker = await nativeWorkerCtor.createReplay(compiledOptions, history);
+    const constructedWorker = await this.bundleWorker(compiledOptions, replayWorker);
+
+    const runPromise = constructedWorker.run();
+    runPromise.catch((err) => {
+      console.error('Caught error while replay worker was running', err);
```
Comment (bergundy) [inline packages/worker/src/worker.ts:222]: Let's not log directly to console

Comment (bergundy) [reply packages/worker/src/worker.ts:222]: We probably shouldn't catch this at all since we await it anyways

File: packages/worker/src/worker.ts [hunk 16]
```diff
@@ -512,16 +562,16 @@ export class Worker {
...
-            filter((state) => state === 'DRAINING'),
+            filter((state) => state === 'SHUTDOWN'),
             first(),
             map((): ContextAware<{ activation: coresdk.workflow_activation.WorkflowActivation; synthetic: true }> => {
               return {
                 parentSpan: this.tracer.startSpan('workflow.shutdown.evict'),
                 activation: new coresdk.workflow_activation.WorkflowActivation({
                   runId: group$.key,
-                  jobs: [{ removeFromCache: { reason: 'Shutting down' } }],
+                  jobs: [{ removeFromCache: { message: 'Shutting down' } }],
```
Comment (bergundy) [inline packages/worker/src/worker.ts:574]: we should probably attach the reason enum here too

File: packages/worker/src/worker.ts (hunk 8 above)
Comment (bergundy) [reply packages/worker/src/worker.ts:817]: I don't know if this is possible, we would probably need to make shutdown error not an `Error`, maybe instead we can return an optional value and break the loop if it's undefined.

File: packages/test/src/test-replay.ts [hunk 17]
```diff
@@ -13,23 +13,25 @@ export interface Context {
 const test = anyTest as TestInterface<Context>;
 
 test.before(async (t) => {
+  // We don't want AVA to whine about unhandled rejections thrown by workflows
+  process.removeAllListeners('unhandledRejection');
   const logger = new DefaultLogger('DEBUG');
   // Use forwarded logging from core
-  const replayCore = await ReplayCore.install({ logger, telemetryOptions: { logForwardingLevel: 'DEBUG' } });
+  const replayCore = await ReplayCore.install({ logger, telemetryOptions: { logForwardingLevel: 'INFO' } });
```
Comment (bergundy) [inline packages/test/src/test-replay.ts:20]: Do we need to configure the tracing filter too?

File: packages/test/tsconfig.json [hunk 18]
```diff
@@ -13,6 +13,6 @@
     { "path": "../worker" },
     { "path": "../workflow" }
   ],
-  "include": ["./src/**/*.ts", "./src/history_files/*.json"],
+  "include": ["./src/**/*.ts", "./history_files/*.json"],
```
Comment (bergundy) [inline packages/test/tsconfig.json:16]: Oh now I see why you put this in `src`, just so you can import JSONs directly...
We don't have any json histories AFAICT so we could probably just remove this line.
I don't like using sources outside of the `src` directory because it means that anything that was in `src` will be outputted to `lib/src`

File: packages/worker/src/core.ts [hunk 19]
```diff
@@ -15,9 +15,9 @@ import { compileServerOptions, getDefaultServerOptions, RequiredServerOptions, S
 import { DefaultLogger, LogEntry, Logger, LogTimestamp, timeOfDayToBigint } from './logger';
 import * as errors from './errors';
 import { temporal } from '@temporalio/proto';
-import History = temporal.api.history.v1.History;
 import { byteArrayToBuffer } from './utils';
 import { MakeOptional } from '@temporalio/common/src/type-helpers';
+import History = temporal.api.history.v1.History;
```
Comment (bergundy) [inline packages/worker/src/core.ts:20]: Let's export this as the interface instead:
```ts
export type History = temporal.api.history.v1.IHistory;
```

File: packages/worker/src/core.ts [hunk 20]
```diff
@@ -291,24 +284,15 @@ export class Core {
...
-  }
-
-  protected static setInstance(i: Promise<ReplayCore>): void {
-    this._replayInstance = i;
-  }
-
-  protected static clearInstance(): void {
-    delete this._replayInstance;
-  }
+  // protected static _instanceCtor = (options: native.CoreOptions, callback: native.CoreCallback): void => {
```
Comment (bergundy) [inline packages/worker/src/core.ts:287]: letftover comment

File: packages/worker/src/worker.ts [hunk 21]
```diff
@@ -229,17 +229,26 @@ export class Worker {
...
+        constructedWorker.shutdown();
+      } catch {
+        // We may have already called shutdown from the eviction branch, so ignore
+        // errors if we're already shut down.
+      }
     });
 
-    const nondeterminismPromise = firstValueFrom(
-      constructedWorker.evictions.pipe(
+    const evcitionPromise = firstValueFrom(
```
Comment (bergundy) [inline packages/worker/src/worker.ts:242]:
```suggestion
    const evictionPromise = firstValueFrom(
```

File: packages/worker/src/worker.ts [hunk 22]
```diff
@@ -229,17 +229,26 @@ export class Worker {
     const constructedWorker = await this.bundleWorker(compiledOptions, replayWorker);
 
     const runPromise = constructedWorker.run();
-    const pollerShutdown = constructedWorker.workflowPollerStateSubject.pipe(
-      filter((state): state is 'SHUTDOWN' => state === 'SHUTDOWN')
-    );
+    const pollerShutdown = constructedWorker.workflowPollerStateSubject.pipe(filter((state) => state !== 'POLLING'));
     pollerShutdown.subscribe((_) => {
```
Comment (bergundy) [inline packages/worker/src/worker.ts:233]: Let's use `firstValueFrom` instead of subscribing.
If you subscribe you need to unsubscribe too which is more error prone (e.g. I missed it in the first 2 reviews).
https://rxjs.dev/api/index/class/Observable#cancel-a-subscription

**Review by bergundy — COMMENTED** [review]
Small leftovers but we're mostly there

File: packages/test/src/test-replay.ts (hunk 17 above)
Comment (Sushisource) [reply packages/test/src/test-replay.ts:20]: Nah, no need. It's easy enough to just manually tweak when desired during test

File: packages/worker/src/core.ts (hunk 19 above)
Comment (Sushisource) [reply packages/worker/src/core.ts:20]: I don't think so - it should be the actual reified type, which ensures that it's actually been parsed properly and doesn't have undefined fields that can't actually be that way in a real history

**Review by bergundy — APPROVED** [review]
